contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,212 @@
+  * I should be able use that device without writing custom Kubernetes code.
+  * I want some mechanism to give priority among pods for particular devices on my node.
+
+## Objectives
+
+1. Create a plugin mechanism which allows discovery and monitoring of devices
+2. Change the API to support RessourceClass and ResourceType
+3. Add support for ResourceType and ResourceType in the scheduler and kubelet
+
+## Non Objectives
+
+1. Solving the runtime is out of scope
+
+## Proposed changes
+
+### API Changes
+#### ResourceType
+
+When discovering the devices, Kubelet will be in charge of advertising those
+resources to the API server.
+
+We will advertise each device returned by the Device Plugin in a new structure
+called the ResourceType.
+It is defined as follows:
+
+```golang
+type ResourceType struct {
+	Kind string
+	Name string
+	Quantity resource.Quantity
+	Properties []string
+}
+```
+
+Because the current API (Capacity, Allocatable) can not be extended to
+support ResourceType, we will need to create two new attributes in the NodeStatus
+structure: `CapacityV2` and `AllocatabeV2`:
+
+```golang
+type NodeStatus struct {
+	Capacity ResourceList
+	CapacityV2 []ResourceType
+
+	Allocatable ResourceList
+	AllocatableV2 []ResourceType
+}
+```
+#### ResourceClass
+
+We will not allow the end user to specify `ResourceTypes` in his pod spec. Instead
+we propose a spec called `ResourceClass` which encapsulate devices requests.
+
+Only cluster admins will be able to define a ResourceClass by default. This can
+then be tweaked through RBAC.
+
+A ResourceClass spec will have the following spec:
+```yaml
+kind: ResourceClass
+apiVersion: compute.k8s.io/v1alpha1
+metadata:
+  name: nvidia-tesla-3.7
+spec:
+    resourceSelector:  // required are ORed
+        - required:  // all properties under required are ANDed
+          - key: kind
+            operator: In
+            values: ["nvidia-gpu"]
+          - key: memory
+            operator: Gt
+            values: ["8 Gi"]
+        - required:
+          - key: kind
+            operator: In
+            values: ["nvidia-gpu"]
+          - key: compute-capabilities
+            operator: Gt
+            values: ["3.7"]
+
+```
+
+The user will then be able to specify in his pod Spec requirements a
+Resource class through either Requests or Limits
+```yaml
+kind: Pod
+apiVersion: v1
+metadata:
+  name: gpu-pod
+spec:
+  containers:
+    name: gpu-container
+    resources:
+      limits:
+        cpu: 1.2
+        memory: 10Gi
+        nvidia-tesla-3.7: 4
+      requests:
+        cpu: 0.3
+```
+
+
+### Device Plugin
+
+Kubelet will create a gRPC connection (as a client) to the Device Plugins whose endpoint
+were specified on the CLI.
+
+The kubelet will call the `Discover` function once and then listen to a stream returned
+by the `Monitor` function.
+
+When receiving a pod which requests GPUs (through the ResourceClass api) kubelet will
```

Comment:
ResourceClasses provide better scheduling primitives than what is available today.
With resource classes, you can request a GPU with specific attributes. Without that you will get any GPU. 
The hardware plugin interface should ideally work with the current resource APIs even if it were to expose additional attributed to enabled ResourceClasses in the future.
Once ResourceClasses are introduced, better scheduling primitives will be available through the same hardware plugin APIs.